x/gamm/keeper/taker_fee.go

```diff
@@ -4,20 +4,111 @@ import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 
 	"github.com/osmosis-labs/osmosis/v15/x/gamm/types"
+	poolmanagertypes "github.com/osmosis-labs/osmosis/v15/x/poolmanager/types"
 )
 
 const (
 	defaultTakerFeeDenom = "udym"
 )
 
-func (k Keeper) chargeTakerFee(ctx sdk.Context, takerFeeCoin sdk.Coin, sender sdk.AccAddress) error {
-	// If the denom is the base denom:
+// This function is a helper function to support the SwapAmountOut msg
+// it finds the route to the defaultTakerFeeDenom to be used to swap the taker fee
+func RouteToBaseDenomFromOutRoutes(routes poolmanagertypes.SwapAmountOutRoutes, denomOut string) []poolmanagertypes.SwapAmountInRoute {
+	var newRoutes []poolmanagertypes.SwapAmountInRoute
+
+	//if the denomOut is the defaultTakerFeeDenom, we need to swap all the routes
+	if denomOut == defaultTakerFeeDenom {
+		for i, route := range routes[:len(routes)-1] {
+			newRoutes = append(newRoutes, poolmanagertypes.SwapAmountInRoute{
+				PoolId:        route.PoolId,
+				TokenOutDenom: routes[i+1].TokenInDenom,
+			})
+		}
+		newRoutes = append(newRoutes, poolmanagertypes.SwapAmountInRoute{
+			PoolId:        routes[len(routes)-1].PoolId,
+			TokenOutDenom: denomOut,
+		})
+		return newRoutes
+	}
+
+	var found bool
+	var idx int
+	// check where it swapped to defaultTakerFeeDenom on the routes
+	for i, route := range routes {
+		if route.TokenInDenom == defaultTakerFeeDenom {
+			found = true
+			idx = i
+			break
+		}
+	}
+
+	//if not found, return empty as we can't swap to the defaultTakerFeeDenom
+	if idx == 0 || !found {
+		return newRoutes
+	}
+
+	for i, route := range routes[:idx] {
+		newRoutes = append(newRoutes, poolmanagertypes.SwapAmountInRoute{
+			PoolId:        route.PoolId,
+			TokenOutDenom: routes[i+1].TokenInDenom,
+		})
+	}
+
+	return newRoutes
+}
+
+func (k Keeper) chargeTakerFeeSwapAmountOut(ctx sdk.Context, takerFeeCoin sdk.Coin, sender sdk.AccAddress, outRoutes []poolmanagertypes.SwapAmountOutRoute, denomOut string) error {
+	if takerFeeCoin.Denom == defaultTakerFeeDenom {
+		return k.burnTakerFee(ctx, takerFeeCoin, sender)
+	}
+
+	routes := RouteToBaseDenomFromOutRoutes(outRoutes, denomOut)
+	if len(routes) == 0 {
+		ctx.Logger().Error("failed to swap taker fee to base denom")
+		return k.communityPoolKeeper.FundCommunityPool(ctx, sdk.NewCoins(takerFeeCoin), sender)
+	}
+
+	err := k.swapAndBurn(ctx, sender, routes, takerFeeCoin)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k Keeper) chargeTakerFeeSwapAmountIn(ctx sdk.Context, takerFeeCoin sdk.Coin, sender sdk.AccAddress, routes []poolmanagertypes.SwapAmountInRoute) error {
 	if takerFeeCoin.Denom == defaultTakerFeeDenom {
 		return k.burnTakerFee(ctx, takerFeeCoin, sender)
-	} else {
-		//TODO: Swap to DYM and burn. will be handled in the future.
+	}
+
+	//build new subroute to swap takerFeeCoin to base denom
+	var newRoutes []poolmanagertypes.SwapAmountInRoute
+	for i, route := range routes {
```

Comment:
seems like the routes are coming here from the routes used to make the switch. but why do we assume that if those routes don't provide a way to dym there isn't other routes that do? seems like we're not checking all routes possible.

e.g the switch USDC -> UDST, which may have a  direct route USDC -> USDT, doesn't mean there isnt UDSC -> X -> DYM - but if I understand correctly the routes for this won't be contained in the check.

Reply:
there's no definitive requirement regarding it.
we don't manage this route mapping, and I'm not sure it's a good idea to randomly pick pools to swap the taker fee.

The issue is that the requirement is to allow pools with different base assets (DYM / USDC) but the taker fee has specific logic for DYM.
if we could treat base assets equally, there will be less edge cases

anyway, more holistic approach will probably be allowing gas with different tokens, then swap them to DYM in gov decided pools. something like that i think

Reply:
> there's no definitive requirement regarding it.
from my understanding of the requirement, every non-dym taker fee should be swapped to dym assuming there is some route which can facilitate this swap. 

> we don't manage this route mapping, and I'm not sure it's a good idea to randomly pick pools to swap the taker fee.
I believe this can be done similar to how the client calculates the route. or alternatively this client can also calculate if there is a route to swap to dym and pass it also in the request.